### ISSUE TITLE
Long input series makes oom

### ISSUE THREAD

**Issue by seanxuu** [datamllab/LongLM#10]
Great job! However, I have encountered an issue. When the length of the input increases, the GPU memory consumption grows rapidly, and it quickly leads to an out-of-memory (OOM) error. Could you please let me know if this is a bug?

**Comment by Mooler0410** [comment]
Current implementation does not consider any efficiency issues. It will nearly double the memory requirement. Our flash attention version will come soon!

**Comment by kungfu-eric** [comment]
This works across 2x48GB GPUs with the mistral example: 

`model_kwargs={"torch_dtype": torch.float16, "load_in_4bit": True},
model = AutoModelForCausalLM.from_pretrained(model_path, config=config, torch_dtype= torch.float16, load_in_4bit= True, device_map="auto")`

Requires bits and bytes installed.

**Comment by kungfu-eric** [comment]
Their 10k example just fits

**Comment by dengxiaotian123** [comment]
> > This works across 2x48GB GPUs with the mistral example:
> > `model = AutoModelForCausalLM.from_pretrained(model_path, config=config, torch_dtype= torch.float16, load_in_4bit= True, device_map="auto")`
> > Requires bits and bytes installed.
> 
> how long can you input？

Have you solved it? I use 2*80GB GPUs. Every time it reaches 13k or above, OOM occurs. The basic model is Baichuan2.

**Comment by Mooler0410** [comment]
You may try Microsoft's deepspeed inference, it can save some memory and time. For 7B model, with deepspeed, you can input sequence > 16k

**Comment by Aniketto16** [comment]
@Mooler0410  
I tried deepspeed inference, but I am getting garbage output when we exceed context length of original model i.e. 4096, can you think what might be the problem ? 

I can share the code if that helps! Thank you!

**Comment by Mooler0410** [comment]
> @Mooler0410  
> 
> I tried deepspeed inference, but I am getting garbage output when we exceed context length of original model i.e. 4096, can you think what might be the problem ? 
> 
> 
> 
> I can share the code if that helps! Thank you!

Of course! We are happy to help you with your scripts.

**Comment by Aniketto16** [comment]
Hello @Mooler0410, 
I am not very well versed in deepspeed but here is what I tried, please correct me where I went wrong? 

```
import math
from transformers import GenerationConfig
import deepspeed
import torch
from torch.nn import DataParallel
from transformers import AutoTokenizer, AutoModelForCausalLM, AutoConfig
from dataclasses import dataclass
from threading import Thread
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    TextIteratorStreamer,
    GenerationConfig,
)

import os
import sys
sys.path.append(os.getcwd())

from supervised_finetuning import get_conv_template
from llama_patch import unplace_flash_attn_with_attn

# Support for long lm : self extend 
import llama_self_extend_patch as LlamaSE
from modify_utils import modify_method_of_instance
from functools import partial
from transformers.models.llama.modeling_llama import LlamaAttention
from transformers import AutoTokenizer, AutoModelForCausalLM

local_rank = int(os.getenv("LOCAL_RANK", "0"))
world_size = int(os.getenv("WORLD_SIZE", "1"))

if __name__ == '__main__':
    
    ctx_length = 8192
    model_path = "./checkpoints/finetune/merged-sft-elyza-fast13b-long"
    quantization = "int8"
    template_name = "llama2"

    config = AutoConfig.from_pretrained(
        model_path
    )

    orig_ctx_len = getattr(config, "max_position_embeddings", None)
    if orig_ctx_len and ctx_length > orig_ctx_len:
        print('Utilizing SelfExtend...')
        original_llama_forward = LlamaAttention.forward
        self_extend_forward = partial(LlamaSE.self_extend_forward, group_size_1=5, group_size_2=1024)
    
    tokenizer = AutoTokenizer.from_pretrained(model_path, 
                                            model_max_length=ctx_length if ctx_length > orig_ctx_len else orig_ctx_len,
                                            trust_remote_code=True)
            
    base_model = AutoModelForCausalLM.from_pretrained(
        model_path,
        torch_dtype=torch.float16,
        low_cpu_mem_usage=True,
        trust_remote_code=True,
    )

    try:
        base_model.generation_config = GenerationConfig.from_pretrained(
            model_path, trust_remote_code=True
        )
    except OSError:
        print("Failed to load generation config, use default.")

    ds_engine = deepspeed.init_inference(
        base_model,
        mp_size=world_size,
        dtype=torch.int8,
    )

    base_model = ds_engine.module
    base_model.eval()

    template = get_conv_template(template_name)
    stop_str = tokenizer.eos_token if tokenizer.eos_token else template.stop_str
    
    #  Prompt starts here --->
    material = read_file('./data/company_data/06.SalesProcessGuidance/フロントセールスプロセスガイダンス.pdf')[:10000]
    
    # system_prompt = template.system_prompt
    system_prompt = """
            - You are a helpful Assistant from company : 大塚商会 who answers to users questions based on given context given to you.
            - Keep your answer detailed but to the point.
            - The content given below is from important company document.
            - Reply "Not applicable" if text is irrelevant to the context.
            - Give markdown formatted response with relevant font stylings.

            content : {con}

            Follow the instructions explicitly as mentioned and do not deviate from the above instructions.
        """
    system_prompt = system_prompt.format(con = material)
    history = [["Can you summarize this document in 10 points for me ?",""]]

    prompt = template.get_prompt(messages=history, system_prompt=system_prompt)
    input_ids = tokenizer(prompt).input_ids
    

    #  Generation starts here --->
    max_new_tokens = 1024
    temperature = 0
    repetition_penalty = 1.125
    do_print = False

    streamer = TextIteratorStreamer(
        tokenizer, timeout=60.0, skip_prompt=True, skip_special_tokens=False
    )

    max_src_len = ctx_length - max_new_tokens - 8
    # max_src_len = context_len
    input_ids = input_ids[-max_src_len:]

    print('context length : ', len(input_ids))

    generation_kwargs = dict(
        input_ids=torch.as_tensor([input_ids]).to(local_rank),
        max_new_tokens=max_new_tokens,
        temperature=temperature,
        repetition_penalty=repetition_penalty,
        streamer=streamer,
        do_sample= True if temperature > 0 else False
    )
    thread = Thread(target=base_model.generate, kwargs=generation_kwargs)
    thread.start()

    generated_text = ""
    for new_text in streamer:
        stop = False
        pos = new_text.find(stop_str)
        if pos != -1:
            new_text = new_text[:pos]
            stop = True
        generated_text += new_text
        if do_print:
            print(new_text, end="", flush=True)
        if stop:
            break
    if do_print:
        print()

    print(generated_text)
```

You can ignore the read_fiile() function it is just to read the document. On the other note, I have referred to this documentation for implementing this : https://www.deepspeed.ai/tutorials/automatic-tensor-parallelism/

I think the reason it is going out of memory is due to the fact that the model is unable to allocate such a large tensor in the GPU memory so I thought tensor parallelism would be the way out! I have 4 x NVIDIA RTX 6000 Ada 50GB GPUs btw... 

Anyways,  thank you so much in advance!

**Comment by Mooler0410** [comment]
Hi ! Seem that you forget to use modify_method_of_instance to bind the self_extend_forward to your model. After setting self_extend_forward, you should call modify_method_of_instance as it in the llama example so as to replace the original llama's forward function. 

After doing this, you should be able to get some reasonable outputs.

Beside, one minor issue may be that: if you want to input a 16k sequence, you'd better set the group size larger than 8 instead of 5 (which you are using now). The reason has been discussed in our paper. You may have a check. TLDR: LLMs' larger position are not well trained.

**Comment by Aniketto16** [comment]
Hello @Mooler0410 ! 
I changed the code to following, still I am getting empty output for 8192 and 16834 context length : 

```
local_rank = int(os.getenv("LOCAL_RANK", "0"))
world_size = int(os.getenv("WORLD_SIZE", "1"))

if __name__ == '__main__':
    
    ctx_length = 8192
    model_path = "./checkpoints/finetune/merged-sft-elyza-fast13b-long"
    quantization = "int8"
    template_name = "llama2"

    # Set RoPE scaling factor
    config = AutoConfig.from_pretrained(
        model_path
    )

    orig_ctx_len = getattr(config, "max_position_embeddings", None)
    if orig_ctx_len and ctx_length > orig_ctx_len:
        # scaling_factor = float(math.ceil(ctx_length / orig_ctx_len))
        # config.rope_scaling = {"type": "linear", "factor": scaling_factor}
        print('Utilizing SelfExtend...')
        original_llama_forward = LlamaAttention.forward
        self_extend_forward = partial(LlamaSE.self_extend_forward, group_size_1=8, group_size_2=1024)
    
    tokenizer = AutoTokenizer.from_pretrained(model_path, 
                                            model_max_length=ctx_length if ctx_length > orig_ctx_len else orig_ctx_len,
                                            trust_remote_code=True)
            
    base_model = AutoModelForCausalLM.from_pretrained(
        model_path,
        # load_in_8bit=True if quantization == "int8" else False,
        # load_in_4bit=True if quantization == "int4" else False,
        # config=config,
        torch_dtype=torch.float16,
        low_cpu_mem_usage=True,
        # device_map={"", local_rank},
        trust_remote_code=True,
    )

    if orig_ctx_len and ctx_length > orig_ctx_len: modify_method_of_instance(base_model, "LlamaAttention", "forward", self_extend_forward)

    try:
        base_model.generation_config = GenerationConfig.from_pretrained(
            model_path, trust_remote_code=True
        )
    except OSError:
        print("Failed to load generation config, use default.")

    ds_engine = deepspeed.init_inference(
        base_model,
        mp_size=world_size,
        dtype=torch.int8,
    )

    base_model = ds_engine.module
    base_model.eval
```

Thank you for your help in advance!

**Comment by Mooler0410** [comment]
You will need to call     "modify_method_of_instance"
To replace the model's default forward function with self-extend.

E.g.     modify_method_of_instance(model, "MistralAttention", "forward", self_extend_forward)

You may check the example codes of how to use selfextend in llama/mistral.

**Comment by Aniketto16** [comment]
@Mooler0410, sorry for unclear code! 

I made the same change in this line : 
```
if orig_ctx_len and ctx_length > orig_ctx_len: modify_method_of_instance(base_model, "LlamaAttention", "forward", self_extend_forward)
```

Is this the same thing you wanted ? 
Thanks

**Comment by Mooler0410** [comment]
Hi! We may need to run your scripts to have a check. Are you using a variant of the 'elyza-fast13b' model on hugging face? If so, we will use that model as a replacement of your local model.

**Comment by Aniketto16** [comment]
@Mooler0410, Yes you can use the instruct variant on the huggingface! 
Here is the link : https://huggingface.co/elyza/ELYZA-japanese-Llama-2-13b-fast-instruct
Thank you again so much in advance!

**Comment by Mooler0410** [comment]
Thanks for the confirmation. We will test your scripts this week!

**Comment by tginart** [comment]
> Current implementation does not consider any efficiency issues. It will nearly double the memory requirement. Our flash attention version will come soon!

Any timeline on a flash attention version? Perhaps using torch's sdpa?

**Comment by Mooler0410** [comment]
> @Mooler0410, Yes you can use the instruct variant on the huggingface! Here is the link : https://huggingface.co/elyza/ELYZA-japanese-Llama-2-13b-fast-instruct Thank you again so much in advance!

Hi Aniketto16!

It works well in our testing.  May I ask you for the version for your transformers?

(Sorry for the late reply, we are busy with the flash-attention implementation and debugging staffs these days)

**Comment by Mooler0410** [comment]
To be more clear, we just found that after 4.36, the default attention of llama change from "LlamaAttention" to "LlamaSdpaAttention". Hence the replacement function will take no effect. This might be the reason for the failure. We will modify the codes to handle such cases!

**Comment by Aniketto16** [comment]
Hi @Mooler0410, 
It is working now, that was the only error, I am really sorry I was not aware of these changes!! 
Thank you so much for your help and I am really sorry for constantly bothering you! 
Thank you!!

**Comment by Aniketto16** [comment]
I think you can close the issue now!
